cpp/src/c_api/uniform_neighbor_sampling.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cugraph_c/algorithms.h>
+
+#include <c_api/abstract_functor.hpp>
+#include <c_api/graph.hpp>
+#include <c_api/utils.hpp>
+
+#include <cugraph/algorithms.hpp>
+#include <cugraph/detail/utility_wrappers.hpp>
+#include <cugraph/graph_functions.hpp>
+#include <cugraph/visitors/generic_cascaded_dispatch.hpp>
+
+#include <raft/handle.hpp>
+
+namespace cugraph {
+namespace c_api {
+
+struct cugraph_sample_result_t {
+  cugraph_type_erased_device_array_t* src_{nullptr};
+  cugraph_type_erased_device_array_t* dst_{nullptr};
+  cugraph_type_erased_device_array_t* label_{nullptr};
+  cugraph_type_erased_device_array_t* index_{nullptr};
+  cugraph_type_erased_host_array_t* count_{nullptr};
+};
+
+}  // namespace c_api
+}  // namespace cugraph
+
+namespace {
+
+#if 0
+// FIXME: ifdef this out for now.  Can't be implemented until PR 2073 is merged
+
+  struct uniform_neighbor_sampling_functor : public cugraph::c_api::abstract_functor {
+  raft::handle_t const& handle_;
+  cugraph_graph_t* graph_{nullptr};
+  cugraph_type_erased_device_array_view_t const* start_{nullptr};
+  cugraph_type_erased_device_array_view_t const* start_label_{nullptr};
+  cugraph_type_erased_host_array_view_t const* fan_out_{nullptr};
+  bool with_replacement_{false};
+  cugraph_sample_result_t* result_{nullptr};
+
+  uniform_neighbor_sampling_functor(::cugraph_resource_handle_t const* handle,
+                                    ::cugraph_graph_t* graph,
+                                    ::cugraph_type_erased_device_array_view_t const* start,
+                                    ::cugraph_type_erased_device_array_view_t const* start_label,
+                                    ::cugraph_type_erased_host_array_view_t const* fan_out,
+                                    bool without_replacement)
```

Comment:
The C++ API defines the semantic of this flag as `with_replacement` (which defaults to `true`). I know this code is `ifdef`-out for now, but it's meant to be brought back, as far as I understand.

Reply:
Yes.  I defined this based on the old definition in the C++ API.  Fixed in latest push.